manager/src/main/java/org/openremote/manager/event/ClientEventService.java

```diff
@@ -162,68 +173,92 @@ public void configure() throws Exception {
                             .to(GatewayService.GATEWAY_EVENT_TOPIC)
                         .endChoice()
                         .stop()
+                    .when(bodyAs(String.class).startsWith(EventSubscription.SUBSCRIBE_MESSAGE_PREFIX))
+                        .process(exchange -> exchange.getIn().setHeader(HEADER_CONNECTION_TYPE, HEADER_CONNECTION_TYPE_WEBSOCKET))
+                        .to(ClientEventService.CLIENT_EVENT_QUEUE)
+                        .stop()
+                    .when(bodyAs(String.class).startsWith(CancelEventSubscription.MESSAGE_PREFIX))
+                        .to(ClientEventService.CLIENT_EVENT_QUEUE)
+                        .stop()
+                    .when(bodyAs(String.class).startsWith(RenewEventSubscriptions.MESSAGE_PREFIX))
+                        .to(ClientEventService.CLIENT_EVENT_QUEUE)
+                        .stop()
+                    .when(bodyAs(String.class).startsWith(SharedEvent.MESSAGE_PREFIX))
+                        .convertBodyTo(SharedEvent.class)
+                        .process(exchange -> {
+                            SharedEvent event = exchange.getIn().getBody(SharedEvent.class);
+                            // If there is no timestamp in event, set to system time
+                            if (event.getTimestamp() <= 0) {
+                                event.setTimestamp(timerService.getCurrentTimeMillis());
+                            }
+                        })
+                        .choice()
+                            .when(exchange -> isGatewayClientId(getClientId(exchange)))
+                                .to(GatewayService.GATEWAY_EVENT_TOPIC)
+                            .otherwise()
+                                .to(ClientEventService.CLIENT_EVENT_TOPIC)
+                        .endChoice()
+                    .otherwise()
+                        .process(exchange -> LOG.fine("Unsupported message body: " + exchange.getIn().getBody()))
+                    .end();
+
+                from(ClientEventService.CLIENT_EVENT_QUEUE)
+                    .routeId("ToClientWebsocketEvents")
+                    .choice()
                     .when(bodyAs(String.class).startsWith(EventSubscription.SUBSCRIBE_MESSAGE_PREFIX))
                         .convertBodyTo(EventSubscription.class)
                         .process(exchange -> {
                             String sessionKey = getSessionKey(exchange);
                             EventSubscription subscription = exchange.getIn().getBody(EventSubscription.class);
                             AuthContext authContext = exchange.getIn().getHeader(Constants.AUTH_CONTEXT, AuthContext.class);
+                            sessionKeyConnectionTypeMap.put(sessionKey, (String) exchange.getIn().getHeader(HEADER_CONNECTION_TYPE));
```

Comment:
Shouldn't be adding/removing sessionKeys for subsribe/unsubscribe messages...should only do this on connect/disconnect

Reply:
A single client could have multiple subscriptions